library/src/scripts/flyouts/NewPostMenu.tsx

```diff
@@ -59,47 +71,285 @@ function ActionItem({ item, style }: { item: IAddPost; style: ITransition }) {
                     {contents}
                 </LinkAsButton>
             )}
-        </div>
+        </animated.li>
     );
 }
 
+interface IState {
+    open: boolean;
+    buttonFocus: boolean;
+    focusedItem?: number;
+}
+
+const initialState: IState = {
+    open: false,
+    buttonFocus: false,
+    focusedItem: undefined,
+};
+
+const reducer = (state, action) => {
+    switch (action.type) {
+        case "toggle_open":
+            return { ...state, open: !state.open };
+        case "set_open":
+            return { ...state, open: action.open };
+        case "set_button_focus":
+            return { ...state, buttonFocus: action.focus };
+        case "set_focused_item":
+            return { ...state, focusedItem: action.item };
+        default:
+            throw new Error();
+    }
+};
+
 export default function NewPostMenu(props: { items: IAddPost[] }) {
-    const [open, setOpen] = useState(false);
-    const toggle = () => setOpen(!open);
-    const onClickBackground = () => {
-        if (open) {
-            setOpen(false);
+    const classes = newPostMenuClasses();
+    const vars = newPostMenuVariables();
+
+    let { items } = props;
+
+    const buttonRef = useRef<HTMLButtonElement>(null);
+    const backgroundRef = useRef<HTMLElement>(null);
+    const menuRef = useRef<HTMLUListElement>(null);
+    const itemsRef = useRef<HTMLLIElement>(null);
+    const accessMenuRef = useRef<HTMLUListElement>(null);
+
+    const [state, dispatch] = useReducer(reducer, initialState);
+
+    const onClickBackground = (event: React.MouseEvent) => {
+        event.stopPropagation();
+        if (state.open) {
+            dispatch({ type: "set_button_focus", focus: true });
+            dispatch({ type: "set_open", open: false });
+            dispatch({ type: "set_focused_item", item: undefined });
         }
     };
 
-    const classes = newPostMenuClasses();
-    const { items } = props;
+    const onBgKeyDown = (event: React.KeyboardEvent<any>) => {
+        event.preventDefault();
+        event.stopPropagation();
+
+        switch (event.key) {
+            case "Escape":
+                if (state.open) {
+                    dispatch({ type: "set_open", open: false });
+                    dispatch({ type: "set_button_focus", focus: true });
+                    dispatch({ type: "set_focused_item", item: undefined });
+                }
+                break;
+            default:
+                // We don't want to lose focus in case we already have
+                if (state.buttonFocus) {
+                    dispatch({ type: "set_button_focus", focus: true });
+                }
+                break;
+        }
+    };
+
+    const onKeyDown = (event: React.KeyboardEvent<any>) => {
```

Comment:
Can we call this "onToggleKeyDown" or something similar to differentiate it from the other handlers? Feel free to also break up this file into smaller components if it's getting too big.

Reply:
Ok, will change.